dm/docs/RFCS/20211130_enhanced_pre_checker.md

```diff
@@ -0,0 +1,83 @@
+
+## Background
+
+Before the DM’s task starts, we should check some items to avoid start-task failures. You can see the details in our [document](https://docs.pingcap.com/tidb-data-migration/stable/precheck#disable-checking-items). In order to allow some users to use DM normally under certain circumstances, we can also set ignore-check-items in task config to ignore some items that you don’t want to check. Now, we find some shortcomings in regard to check-items.
+
+### Bad user habits
+
+We allow users to ignore all check-items, in which case the user's privilege is too large to perform unexpected operations. 
+
+### Too much time overhead
+
+If we have a large number of tables in source, we will take too much time in checking table schema, sharding table consistency and sharding table auto increment ID.
+
+### Inadequate check
+
+* Now we check auto increment ID by `column-mapping` which is deprecated. If we don't set the `column-mapping` and don't ignore the **auto_increment_ID**, the pre-check will report an error. PS: **auto_increment_ID** is only checked when **schema_of_shard_tables** is true.
+* Dump privilege only checks RELOAD and SELECT. However, Dumpling supports different [consistency configurations](https://docs.pingcap.com/tidb/stable/dumpling-overview#adjust-dumplings-data-consistency-options), which need more privileges.
+* If online-ddl is set by true and a DDL is in online-ddl stage, DM will have a problem in all mode. Specifically, when ghost table has been created but not been renamed, DM will report an error when the ghost table is renamed during the incremental phase. You can learn more about online-ddl [here](https://docs.pingcap.com/tidb-data-migration/stable/feature-online-ddl).
+* For schema_of_shard_tables, whether it's a pessimistic task or an optimistic task, we all simply check consistency of schema by comparing all sharding tables’ structures. For optimistic mode, we can do better.
+
+## Proposal
+
+### Speed ​​up check
+
+1. Support concurrent check
+    - table_schema
+    - schema_of_shard_tables
+    - auto_increment_ID
+2. We casn adjust the concurrency by table numbers.
+
+### Optimize some check
+
+1. **auto_increment_ID** only checked in sharding mode.
+    - If table exist auto increment ID, report a warning to user and tell them the method that can resolve the PK/UK conflict;
+        1. If you set PK to AUTO_INCREMENT, you must make sure that the primary key in sharding tables is not duplicated;
+        2. If sharding tables have duplicated PK, you can refer to [document](https://docs.pingcap.com/tidb-data-migration/stable/shard-merge-best-practices#handle-conflicts-of-auto-increment-primary-key).
+2. Dump_privilege will check different privileges according to different [consistency](https://docs.pingcap.com/tidb/stable/dumpling-overview#adjust-dumplings-data-consistency-options) and downstream on source.
+    - For all consistency, we will check
+        - REPLICATION CLIENT (global)
```

Comment:
this is used in incremental phase, I think dumpling don't need it.

Also, all consistency options should be aligned with https://github.com/pingcap/ticdc/pull/3674#discussion_r762777211

Reply:
ptal @lichunzhu 

Why dumpling need REPLICATION CLIENT privilege?

Reply:
Because `SHOW MASTER STATUS` needs `REPLICATION CLIENT` privilege?

Reply:
also `SHOW SLAVE HOSTS` need `REPLICATION SLAVE` privilege

https://github.com/pingcap/ticdc/blob/84acdbc550c66cf28ee2412afbe6e07fbe39ac84/dm/pkg/utils/db.go#L106

Reply:
DM has a test of revoking REPLICATION CLIENT and start a full mode task. So I expect dumpling will not fail the dump without this privilege and we only require it for incremental task

Reply:
After test, dump phase does not need REPLICATION CLIENT and REPLICATION SLAVE. I will remove them.

Reply:
Maybe we should check this privilege according to the task mode.
For all/incremental task we need these two privileges. For full mode it's not needed.

Reply:
Yes, we still have ReplicationPrivilegeChecker, which can check them in all/incremental task.